core/src/main/scala/org/apache/spark/executor/ExecutorMetrics.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.executor
+
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.metrics.ExecutorMetricType
+
+/**
+ * :: DeveloperApi ::
+ * Metrics tracked for executors and the driver.
+ *
+ * Executor-level metrics are sent from each executor to the driver as part of the Heartbeat.
+ */
+@DeveloperApi
+class ExecutorMetrics private[spark] extends Serializable {
+
+  // Metrics are indexed by MetricGetter.values
+  private val metrics = new Array[Long](ExecutorMetricType.values.length)
```

Comment:
I suggested this earlier in the reviews.  Most of the operations for dealing with this data want to iterate over all the fields.  its much easier this way vs. having a bazillion

```scala
if (x.fizz > y.fizz) { 
  y.fizz = x.fizz
}
if (x.buzz > y.buzz) {
  y.buzz = x.buzz
}
...
```

Reply:
Yup that's fine - I did some googling, unfortunately there isn't a great way to iterate over fields of a case class. You could create a thin wrapper object around the array instead though, if we really think the nicer API is worthwhile:

```
case class Metrics(values: Seq[Long]) {
  def someMetric1(): Long = values(0)
  def ....
  def ...
}
```

Or even this:

```
case class Metrics(metric1: Long, metric2: Long, metfic3: Long, ...) {
  def values(): Seq[Long] = Seq(metric1, metric2, metric3, ...)
}
```

The latter which would be better because you'd be guaranteed to create the struct with the right number of metrics. Though such abstractions are not necessary by any means.

Reply:
Is it likely that users would want to access the individual fields, rather than iterating through all? The 1st option would be a bit nicer if so.

Reply:
Unclear - if we expose these metrics to some external consumer via an API for example, then we almost certainly want to have a schema labelling these fields for consumption by e.g. dashboards. I think what we have here is fine for now.